src/serialization/interval_log/mod.rs

```diff
@@ -0,0 +1,683 @@
+pub struct Tag<'a>(&'a str);
+
+impl<'a> Tag<'a> {
+    /// Create a new Tag.
+    ///
+    /// If a disallowed character is present, this will return `None`.
+    pub fn new(s: &'a str) -> Option<Tag<'a>> {
+        if s.chars()
+            .any(|c| c == ',' || c == '\r' || c == '\n' || c == ' ')
+        {
+            None
+        } else {
+            Some(Tag(s))
+        }
+    }
+
+    /// Returns the tag contents as a str.
+    pub fn as_str(&self) -> &'a str {
+        self.0
+    }
+}
+
+impl<'a> ops::Deref for Tag<'a> {
+    type Target = str;
+
+    fn deref(&self) -> &Self::Target {
+        self.as_str()
+    }
+}
+
+/// An individual interval histogram parsed from an interval log.
+#[derive(PartialEq, Debug)]
+pub struct IntervalLogHistogram<'a> {
+    tag: Option<Tag<'a>>,
+    start_timestamp: f64,
+    // lazily map to Duration to save parsing time and a few bytes of space on ILH
+    duration: f64,
+    max: f64,
+    encoded_histogram: &'a str,
+}
+
+impl<'a> IntervalLogHistogram<'a> {
+    /// Tag, if any is present.
+    pub fn tag(&self) -> Option<Tag<'a>> {
+        self.tag
+    }
+
+    /// Timestamp of the start of the interval in seconds, expressed as a `Duration` relative to
+    /// some start point.
+    ///
+    /// The timestamp may be absolute vs the epoch, or there may be a `StartTime` or `BaseTime` for
+    /// the log, in which case you may wish to consider this number as a delta vs those timestamps.
+    /// See the module-level documentation about timestamps.
+    pub fn start_timestamp(&self) -> time::Duration {
+        fp_seconds_as_duration(self.start_timestamp)
+    }
+
+    /// Duration of the interval in seconds.
+    pub fn duration(&self) -> time::Duration {
+        fp_seconds_as_duration(self.duration)
+    }
+
+    /// Max value in the encoded histogram
+    ///
+    /// This max value is the max of the histogram divided by some scaling factor (which may be
+    /// 1.0).
+    pub fn max(&self) -> f64 {
+        self.max
+    }
+
+    /// Base64-encoded serialized histogram.
+    ///
+    /// If you need the deserialized histogram, base64-decode and use a `Deserializer` on the
+    /// resulting bytes.
+    pub fn encoded_histogram(&self) -> &'a str {
+        self.encoded_histogram
+    }
+}
+
+#[derive(PartialEq, Debug)]
+/// Represents one non-comment line in an interval log.
+///
+/// One thing to note is that the way your interval timestamps work can vary. If your log was
+/// written with a StartTime or BaseTime, that metadata will appear in header comments, and that
+/// will be represented by the iterator providing the corresponding variants here. The presence
+/// of those timestamps will affect how you should interpret the timestamps for individual
+/// intervals. See the module-level documentation.
+#[allow(variant_size_differences)]
+pub enum LogEntry<'a> {
+    /// Logs may include a StartTime. If present, it represents seconds since the epoch.
+    StartTime(f64),
```

Comment:
And I suppose this should be `Duration` then too. More expensive to parse into `Duration` but these should be rare relative to histograms, so it's probably not an issue.

Reply:
I take it back, that should be `SystemTime` I suppose.

Reply:
Hm, now I'm not so sure. There's not really much useful you can do with a `SystemTime` except turn it back into a `Duration` based on `UNIX_EPOCH`, which would force the user to do another round of error checking on the `Result`. So, it seems probably more desirable to give them the `Duration` directly. The path from `Duration` to `chrono::naive::NaiveDateTime` and friends is then pretty smooth.